// oxygen-mdc/oxy-tab.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { LitElement, css, html } from 'lit';
import { customElement, property } from 'lit/decorators';
/**
 * A simple tab component that can be used in conjuction with <oxy-tabs> to
 * ensure only one tab is active at a time.
 */
let OxyTab = class OxyTab extends LitElement {
    constructor() {
        super();
        this.orientation = 'horizontal';
        this.selected = false;
        this.setAttribute('role', 'tab');
        this.setAttribute('tabindex', '0');
    }
    static get styles() {
        return css `
      :host {
        position: relative;
        display: block;
        box-sizing: border-box;
        cursor: pointer;
        outline: none;
        overflow: hidden;
        min-width: 64px;
        user-select: none;
        opacity: 0.7;
      }
      :host(:hover),
      :host(:focus),
      :host([selected]) {
        opacity: 1.0;
      }
      :host([orientation="horizontal"]) {
        padding: 8px 16px;
      }
      :host([orientation="vertical"]) {
        padding: 8px 16px;
      }
      :host::before {
        display: block;
        content: "";
        position: absolute;
        background-color: var(--oxy-tab-indicator-color, #333);
        transition: var(--oxy-tab-animation-duration, 100ms) all;
        transition-timing-function: cubic-bezier(.15, .35, .5, 1.5);
      }
      :host([orientation="horizontal"])::before {
        bottom: 0;
        left: 50%;
        height: 3px;
        width: 0%;
        border-radius: 3px 3px 0 0;
        box-shadow: 0 -1px 2px rgba(0, 0, 0, 0.0);
      }
      :host([orientation="horizontal"][selected])::before {
        box-shadow: 0 -1px 2px rgba(0, 0, 0, 0.5);
        width: 50%;
        left: 25%;
      }
      :host([orientation="vertical"])::before {
        left: 0;
        bottom: 50%;
        width: 3px;
        height: 0%;
        border-radius: 0 3px 3px 0;
        box-shadow: -1px 0 2px rgba(0, 0, 0, 0.0);
      }
      :host([orientation="vertical"][selected])::before {
        box-shadow: 1px 0 2px rgba(0, 0, 0, 0.5);
        height: 70%;
        bottom: 15%;
      }
    `;
    }
    render() {
        return html `<slot></slot>`;
    }
};
__decorate([
    property({ type: String, reflect: true })
], OxyTab.prototype, "orientation", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], OxyTab.prototype, "selected", void 0);
OxyTab = __decorate([
    customElement('oxy-tab')
], OxyTab);
export { OxyTab };
